columnflow/production/cms/pileup.py

```diff
@@ -18,13 +20,98 @@
 set_ak_column_f32 = functools.partial(set_ak_column, value_type=np.float32)
 
 
+logger = law.logger.get_logger(__name__)
+
+
 @producer(
     uses={"Pileup.nTrueInt"},
     produces={"pu_weight", "pu_weight_minbias_xs_up", "pu_weight_minbias_xs_down"},
     # only run on mc
     mc_only=True,
+    # function to determine the correction file
+    get_pileup_file=(lambda self, external_files: external_files.pu_sf),
 )
 def pu_weight(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Based on the number of primary vertices, assigns each event pileup weights using correctionlib.
+    """
+    # compute the indices for looking up weights
+    indices = events.Pileup.nTrueInt.to_numpy().astype("int32") - 1
+
+    # map the variable names from the corrector to our columns
+    variable_map = {
+        "NumTrueInteractions": indices,
+    }
+
+    for column_name, syst in (
+        ("pu_weight", "nominal"),
+        ("pu_weight_minbias_xs_up", "up"),
+        ("pu_weight_minbias_xs_down", "down"),
+    ):
+        # get the inputs for this type of variation
+        variable_map_syst = dict(variable_map, systematic=syst)
+        inputs = [variable_map_syst[inp.name] for inp in self.pileup_corrector.inputs]
+
+        # evaluate and store the produced column
+        pu_weight = self.pileup_corrector.evaluate(*inputs)
+        events = set_ak_column(events, column_name, pu_weight, value_type=np.float32)
+
+    return events
+
+
+@pu_weight.requires
+def pu_weight_requires(self: Producer, reqs: dict) -> None:
+    """
+    Adds the requirements needed the underlying task to derive the pileup weights into *reqs*.
+    """
+    if "external_files" in reqs:
+        return
+
+    from columnflow.tasks.external import BundleExternalFiles
+    reqs["external_files"] = BundleExternalFiles.req(self.task)
+
+
+@pu_weight.setup
+def pu_weight_setup(
+    self: Producer,
+    reqs: dict,
+    inputs: dict,
+    reader_targets: InsertableDict,
+) -> None:
+    """
+    Loads the pileup calculator from the external files bundle and saves them in the
+    py:attr:`pileup_corrector` attribute for simpler access in the actual callable.
+    """
+    logger.warning(
```

Comment:
I think we can remove the warning. If the producer results in the same weight vector (or even a fixed one in case the previous implementation was buggy), this would require explicit communication.

Reply:
Just a comment from the side line:  the logger warning is the result of a discussion @mafrahm and I had offline. We concluded that it would be good to at least temporarily make users aware that the calculation of the pu weights has changed - we had talked about fading this out from one release to another or maybe leave it in for a couple of months just so that people have the chance to realize that this has changed. Personally, I'm still in favor for this, but I also don't have too strong feelings either. So if we decide to not show this at all, I can also live with that

Reply:
I also have no strong opinion on this. I added that it is still recommended to use this Producer.
My reasoning for this warning was to make people aware of a (probably minor) source of inconsistency in results between versions of Columnflow when using this Producer. But I'd also be fine with removing the warning.

Reply:
Fine by me :+1:
I was mainly worried about this warning showing up too often (e.g. when running the producer locally for tasks with a lot of branches / files). I'd suggest to use `logger.warning_once(message_id: str, message: str)`.

Reply:
I will quickly add that to go ahead with this PR.